lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,667 @@
+		}
+	}
+}
+
+impl InteractiveTxStateMachine<Negotiating> {
+	fn receive_tx_complete(self) -> Result<InteractiveTxStateMachine<TheirTxComplete>, InteractiveTxStateMachine<NegotiationAborted>> {
+		match self.is_current_transaction_state_able_to_complete() {
+			Err(e) => Err(InteractiveTxStateMachine { context: self.context, state: NegotiationAborted(e) }),
+			_ => Ok(InteractiveTxStateMachine {
+				context: self.context,
+				state: TheirTxComplete {}
+			})
+		}
+	}
+
+	fn send_tx_complete(self) -> InteractiveTxStateMachine<OurTxComplete> {
+		// TODO: Should we validate before transitioning states? If so, do we want to abort negotiation
+		// if our current transaction state is invalid?
+		InteractiveTxStateMachine {
+			context: self.context,
+			state: OurTxComplete {}
+		}
+	}
+}
+
+impl InteractiveTxStateMachine<OurTxComplete> {
+	fn receive_tx_complete(self) -> Result<InteractiveTxStateMachine<NegotiationComplete>, InteractiveTxStateMachine<NegotiationAborted>> {
+		match self.is_current_transaction_state_able_to_complete() {
+			Err(e) => Err(InteractiveTxStateMachine { context: self.context, state: NegotiationAborted(e) }),
+			_ => Ok(InteractiveTxStateMachine {
+				context: self.context,
+				state: NegotiationComplete {}
+			})
+		}
+	}
+}
+
+impl InteractiveTxStateMachine<NegotiationComplete> {
+	fn get_psbt(&self) -> Result<Transaction, AbortReason> {
+		// Build transaction from inputs & outputs in `NegotiationContext`.
+		return Ok(Transaction {
+			version: self.context.base_tx.version,
+			lock_time: self.context.base_tx.lock_time,
+			input: self.context.inputs.values().cloned().collect(),
+			output: self.context.outputs.values().cloned().collect(),
+		})
+	}
+}
+
+
+enum ChannelMode {
+	Negotiating(InteractiveTxStateMachine<Negotiating>),
+	OurTxComplete(InteractiveTxStateMachine<OurTxComplete>),
+	TheirTxComplete(InteractiveTxStateMachine<TheirTxComplete>),
+	NegotiationComplete(InteractiveTxStateMachine<NegotiationComplete>),
+	NegotiationAborted(InteractiveTxStateMachine<NegotiationAborted>),
+	Indeterminate,
+}
+
+impl Default for ChannelMode {
+	fn default() -> Self { Indeterminate }
+}
+
+pub(crate) struct InteractiveTxConstructor {
+	mode: ChannelMode,
+}
+
+impl InteractiveTxConstructor {
+	pub(crate) fn new(
+		channel_id: [u8; 32],
+		require_confirmed_inputs: bool,
+		is_initiator: bool,
+		base_tx: Transaction,
+		did_send_tx_signatures: bool,
+	) -> Self {
+		let initial_state_machine = InteractiveTxStateMachine::new(
+			channel_id,
+			require_confirmed_inputs,
+			is_initiator,
+			base_tx,
+			did_send_tx_signatures
+		);
+		Self {
+			mode: ChannelMode::Negotiating(initial_state_machine)
+		}
+	}
+
+	pub(crate) fn abort_negotation(&mut self, reason: AbortReason) {
+		self.handle_negotiating_receive(|state_machine| state_machine.abort_negotiation(reason))
+	}
+
+	pub(crate) fn receive_tx_add_input(&mut self, serial_id: SerialId, transaction_input: TxAddInput, confirmed: bool) {
```

Comment:
@dunxen one thing that hasn't been super clear for me -- do we want `InteractiveTxConstructor`'s methods to take in the messages `TxAddInput/TxAddOutput/TxRemoveInput/TxRemoveOutput...` or `TxIn/TxOut`?

Reply:
We want to take in the full message in my opinion. Just makes things easier and we have all the context (even if currently, some seems unnecessary). Also does TxAddInput not already have the serial_id? Maybe we can update the messages for it to be `SerialId` instead of just `u64`.